File: Bugs.php
<?php
require 'vendor/autoload.php';

 use Goutte\Client;
 use MongoDB\Client as MongoClient;

 $dotenv = Dotenv\Dotenv::createImmutable(__DIR__);
 $dotenv->load();
 $mongoClient = new MongoDB\Client($_ENV['MONGO_CONNECTION_STRING']);

 while(true) {
    // MongoDB에 연결
    $mongoCollection = $mongoClient->Music->Bugs;
    $mongoCollection->deleteMany([]);

    // Goutte Client 생성
    $client = new Client();

    // HTTP 요청 헤더 설정
    $header = [
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36',
    ];

    // Goutte Client의 HTTP 요청 헤더 설정
    $client->setServerParameter('HTTP_USER_AGENT', $header['User-Agent']);

    // 크롤링할 페이지 URL
    $url = "https://music.bugs.co.kr/chart";

    // HTML 페이지 가져오기
    $crawler = $client->request('GET', $url);

    // 크롤링한 데이터를 MongoDB에 저장
    $rank_nodes = $crawler->filter('div.ranking > strong');
    $title_nodes = $crawler->filter('p.title > a');
    $singer_nodes = $crawler->filter('td.left > p.artist > a:nth-child(1)');
    $pattern = '/[^\p{Hangul}\p{Latin}\p{Nd}\s\'\x{2019}&]+/u';

    $documents = array();
    $rank_nodes->each(function ($node, $index) use ($title_nodes, $singer_nodes, &$documents, $pattern) {
        $title_text = $title_nodes->eq($index)->text();
        $title_filtered = preg_replace($pattern, '', $title_text);
        $title_filtered = strtoupper($title_filtered);
        $title_filtered = str_replace('PROD BY', 'PROD', $title_filtered);
        $title_filtered = str_replace('’', '\'', $title_filtered);
        $title_filtered = trim($title_filtered);

        $documents[] = array(
            'rank' => (int) $node->text(),
            'title' => $title_filtered,
            'singer' => $singer_nodes->eq($index)->text(),
        );
    });

    $mongoCollection->insertMany($documents);

    $documents = $mongoCollection->find();
    sleep(3600);
}
?>

File: Bugs.py
import requests
import time
import regex as re  # Using regex package instead of re
import os

from bs4 import BeautifulSoup
from pymongo import MongoClient
from dotenv import load_dotenv

# MongoDB Connection
load_dotenv()
mongo_connection_string = os.getenv('MONGO_CONNECTION_STRING')
mongoClient = MongoClient(mongo_connection_string)

while True:
    # Connect to MongoDB
    mongoCollection = mongoClient['Music']['Melon'] 
    mongoCollection.delete_many({})

    # User-Agent
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36'
    }

    # URL to scrape
    url = "https://music.bugs.co.kr/chart"

    # Request the page
    response = requests.get(url, headers=headers)

    # Parse the page with BeautifulSoup
    soup = BeautifulSoup(response.content, 'html.parser')

    documents = []
    pattern = re.compile(r"[^\p{Hangul}\p{Latin}\p{Nd}\s'\u2019&]+", re.UNICODE)

    rank_nodes = soup.select('div.ranking > strong')
    title_nodes = soup.select('p.title > a')
    singer_nodes = soup.select('td.left > p.artist > a:nth-child(1)')

    for index in range(len(rank_nodes)):
        title_text = title_nodes[index].text
        title_filtered = re.sub(pattern, '', title_text)
        title_filtered = title_filtered.upper().replace('PROD BY', 'PROD').replace('’', "'").strip()

        documents.append({
            'rank': int(rank_nodes[index].text),
            'title': title_filtered,
            'singer': singer_nodes[index].text,
        })

    # Insert into MongoDB
    mongoCollection.insert_many(documents)

    # Sleep for an hour
    time.sleep(3600)
